source/codegen/templates/proto_helpers.mako

```diff
@@ -79,23 +85,38 @@ message ${common_helpers.snake_to_pascal(function)}Request {
 % for parameter in input_parameters:
 <%
   parameter_name = common_helpers.camel_to_snake(parameter["name"])
-  grpc_field_number = proto_helpers.generate_parameter_field_number(parameter, used_indexes)
+  is_array = common_helpers.is_array(parameter["type"])
   if 'grpc_type' in parameter:
     parameter_type = parameter['grpc_type']
   else:
-    parameter_type = proto_helpers.determine_function_parameter_type(parameter, service_class_prefix)
+    parameter_type = proto_helpers.get_grpc_type_from_ivi(parameter["type"], is_array, service_class_prefix)
 %>\
 %   if common_helpers.is_enum(parameter):
 <%
-  is_array = common_helpers.is_array(parameter["type"])
-  non_enum_type = proto_helpers.get_grpc_type_from_ivi(parameter["type"], is_array, service_class_prefix)
+  enum_parameter_type = None
+  mapped_type = None
+  if parameter.get('enum', None):
```

Comment:
These conditionals are really exploding and have some duplicated code.

What about a python helper that returns a list of (type, name, number) as namedtuples? And then loop over them in the mako?

Reply:
I agree, we need to find a way of factoring this out.